src/System.Memory/src/System/Buffers/ArrayMemoryPool.cs

```diff
@@ -0,0 +1,28 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Buffers
+{
+    internal sealed partial class ArrayMemoryPool<T> : MemoryPool<T>
+    {
+        private const int DefaultSize = 4096;
+
+        private static readonly ArrayMemoryPool<T> s_shared = new ArrayMemoryPool<T>();
+        public static new ArrayMemoryPool<T> Shared => s_shared;
+
+        public sealed override int MaxBufferSize => 1024 * 1024 * 1024;
+
+        public sealed override OwnedMemory<T> Rent(int minimumBufferSize = -1)
+        {
+            if (minimumBufferSize == -1)
+                minimumBufferSize = DefaultSize;
+            else if (minimumBufferSize > MaxBufferSize || minimumBufferSize < 1)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumBufferSize);
+
+            return new ArrayMemoryPoolBuffer(minimumBufferSize);
```

Comment:
Every rental allocates?

Reply:
It's not clear to me how that could be avoided. The `OwnedMemory` wrapper object is meant to be used in a "using () {}" type-pattern. If we tried to pool those as well, we'd be handing out IDisposable objects that the caller isn't meant to dispose.

Reply:
I'm not really commenting on the implementation so much as the MemoryPool design.  Seems problematic if the default implementation that most folks will use allocates on every Rent.

Reply:
This depends on what the MemoryPool abstraction is meant to be used for.

I see it designed for large buffers. You get large buffer and allocate a little bit in the process. I do not think that allocation is problematic in this scenario.

Reply:
The default size here is 4096, which is also a common size used with ArrayPool, and for ArrayPool, there have been complaints about things as small as the virtual dispatch cost associated with calling ArrayPool.Shared.Rent/Release; even a small allocation and the associated GC impact is at least as much as that.  I'm skeptical this will only be used for large buffers; of course, if we add costs like such allocations on every Rent/Release, we may certainly force it in that direction.

Reply:
If you want a bare bone malloc/free, you should use ArrayPool. MemoryPool is higher level abstraction.

And yes - we need to fix ArrayPool to not have the unnecessary overheads and to not leak.

Reply:
Still seems like it's going to be a problem.  I guess we'll have to agree to disagree.

Reply:
The memory pool in kestrel pools the `OwnedMemory<byte>` objects 😄

Reply:
You can certainly pool OwnedMemory memory if you know what you are doing. It increases your security risk profile.

Reply:
We used to have a cookie in OM to allow it to be pooled safely. We removed the cookie during the performance push, but I am not sure it actually helps performance, yet it makes it so much more difficult to pool OM safely.